### PR TITLE
feat: add NonEmptyArray type

### PR DESCRIPTION
**Pull request by yss14**
**Changes:**
* add new `NonEmptyArray` type
* add missing readme menu entry for `NonEmptyObject` type

Resolves https://github.com/ts-essentials/ts-essentials/issues/307

### CODE REVIEW

File: README.md
Comment (Beraliv) [inline README.md:676]: Would be cool to have this feature in TypeScript, but it's not implemented yet 😅

```suggestion
type FunctionAcceptingRestParameters = (someString: string, ...args: NonEmptyRestParameters) => void;
```

File: README.md
Comment (Beraliv) [inline README.md:681]: Might be worth to have 2 lines instead of 1:

```suggestion
// @ts-expect-error: Type '[]' is not assignable to type 'NonEmptyArray<number>'. Source has 0 element(s) but target requires 1.
const error: NonEmptyArray<number> = [];
```

WDYT?

File: test/index.ts
Comment (Beraliv) [inline test/index.ts:843]: I know that examples in this file are not well structured, but could you please change it to:

```ts
type cases = [
  Assert<IsExact<NonEmptyArray<T>, [T, ...T[]]>>,
  AssertFalse<IsExact<NonEmptyArray<T>, []>>,
  ....
];
```

I think it contains less types that we don't use anyway. And if you want to highlight it somehow, you can write comment right before the test case

WDYT?

**Review by Beraliv — CHANGES_REQUESTED** [review]
Thank you for the help! LGTM! 👍

There are only several places that I would adjust a little bit before merging it.

File: README.md
Comment (yss14) [reply README.md:676]: Ah damn, I think my original idea was to make NonEmptyRestParameters generic :D

**Comment by yss14** [comment]
> Thank you for the help! LGTM! 👍
> 
> There are only several places that I would adjust a little bit before merging it.

Thanks for your review and I totally agree with your comments and suggestions!
I pushed some new commits which implement the proposals.
Looking forward to a second review from your side :)

File: test/index.ts
Comment (Beraliv) [inline test/index.ts:845]: Thank you for the change and all in all, smart idea with generic type in `Cases<T>`!

File: README.md
Comment (Beraliv) [inline README.md:676]: Not critical comment

You can use it here as is:

```suggestion
type FunctionAcceptingRestParameters = (someString: string, ...args: NonEmptyArray<number>) => void;
```

And then don't forget to remove previous lines

**Review by Beraliv — APPROVED** [review]
Thank you for the changes and all in all LGTM!

File: README.md
Comment (yss14) [reply README.md]: Makes sense to simplify the example here a little bit :)

**Comment by yss14** [comment]
@Beraliv I pushed the changes. If it's all fine now, I guess you must merge the PR, since I'm lacking permissions to do so. But no hurries :)

**Comment by psmolinsky** [comment]
Hi guys, btw wouldn't it be also useful to have a type predicate for the `NonEmptyArray` (or `NonEmptyObject` too)? Example:

```typescript
export function isNonEmpty<T>(array: T[] | null | undefined): array is NonEmptyArray<T>;
export function isNonEmpty<T>(array: readonly T[] | null | undefined): array is Readonly<NonEmptyArray<T>>;
export function isNonEmpty<T>(array: readonly T[] | null | undefined): boolean {
    return Array.isArray(array) && array.length > 0;
}
```

**Comment by yss14** [comment]
@psmolinsky I would think that this might be out of scope since ts-essentials is all about typescript types.

**Comment by psmolinsky** [comment]
One more suggestion: would it be useful to have a readonly version of `NonEmptyArray<T>`? My first guess was that `Readonly<NonEmptyArray<T>>` would do the job but it still has some properties writable e.g. `length`:

```typescript
const array: Readonly<NonEmptyArray<number>> = Object.freeze([1, 2] as const);

array.length = 10; // This compiles but fails in runtime.
```

**Comment by Beraliv** [comment]
> One more suggestion: would it be useful to have a readonly version of `NonEmptyArray<T>`? My first guess was that `Readonly<NonEmptyArray<T>>` would do the job but it still has some properties writable e.g. `length`:
> 
> ```ts
> const array: Readonly<NonEmptyArray<number>> = Object.freeze([1, 2] as const);
> 
> array.length = 10; // This compiles but fails in runtime.
> ```

Hey @psmolinsky !

Thank you for the suggestion!

I think in your case you don't need to have `Readonly<NonEmptyArray<number>>`

If you're worried that you can miss empty array here somehow:

1. You can put your configuration (or array) into a separate file
2. Add a test for your file checking if it's empty or not
3. Use it in other files without checking (then `array.length = <number>` will be prohibited)

Playground – https://tsplay.dev/NDk91w

Did I understand you case correctly or there is something I missed here?

**Comment by Beraliv** [comment]
@yss14 could you please also add changelog with `yarn changelog`?

The example is here: https://github.com/ts-essentials/ts-essentials/commit/6b27ee4b7849d2a2670d80000a928cd20e3981b7

I think it should be `minor` as it's adding a new feature

**Comment by yss14** [comment]
@Beraliv Sure thing! I've added a changeset to reflect the minor change.

**Comment by Beraliv** [comment]
Thank you for your changes, I will merge it now

Sorry for waiting, hope it won't block you anyhow

**Pull request merged**
